test/web/RawSocketHost.cs

```diff
@@ -0,0 +1,184 @@
+using System;
+using System.Diagnostics;
+using System.Net;
+using System.Net.Sockets;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc.ApplicationModels;
+using Microsoft.Extensions.Configuration;
+using Tmds.LinuxAsync.Transport;
+using Socket = Tmds.LinuxAsync.Socket;
+using SocketAsyncEventArgs = Tmds.LinuxAsync.SocketAsyncEventArgs;
+
+namespace web
+{
+    public class RawSocketHost
+    {
+        private const int BufferSize = 512;
+        private const string Response =
+            "HTTP/1.1 200 OK\r\nDate: Tue, 31 Mar 2020 14:49:06 GMT\r\nContent-Type: application/json\r\nServer: Kestrel\r\nContent-Length: 27\r\n\r\n{\"message\":\"Hello, World!\"}";
+
+        private static ReadOnlySpan<byte> RequestEnd => new byte[] {13, 10, 13, 10}; // "\r\n\r\n"
+        
+        private readonly string[] _args;
+        private readonly CommandLineOptions _options;
+
+        private IPEndPoint _serverEndpoint;
+
+        public RawSocketHost(CommandLineOptions options, string[] args)
+        {
+            _options = options;
+            _args = args;
+            
+            ConfigurationBuilder bld = new ConfigurationBuilder();
+            bld.AddCommandLine(_args);
+            var cfg = bld.Build();
+            string url = cfg["urls"];
+
+            string[] data = url.Substring(7).Split(':');
+            IPAddress ip = IPAddress.Parse(data[0]);
+            int port = int.Parse(data[1]);
+
+            _serverEndpoint = new IPEndPoint(ip, port);
+        }
+
+        public void Run()
+        {
+            RunAsync().GetAwaiter().GetResult();
+        }
+
+        public async Task RunAsync()
+        {
+            using Socket listener = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+            listener.Bind(_serverEndpoint);
+            listener.Listen(1024);
+            Console.WriteLine($"Raw server listening on {_serverEndpoint}");
+
+            bool deferSends = _options.DeferSends == true;
+            bool deferReceives = _options.DeferReceives == true;
+            bool runContinuationsAsynchronously =
+                _options.SocketContinuationScheduler == SocketContinuationScheduler.ThreadPool;
+
+            using ClientConnectionHandler clientHandler = 
+                new ClientConnectionHandler(deferSends, deferReceives, runContinuationsAsynchronously);
+            
+            while (true)
+            {
+                using Socket handlerSocket = await listener.AcceptAsync();
+
+                clientHandler.HandleClient(handlerSocket);
```

Comment:
We want to new up a ClientConnectionHandler for each Socket, that way we can handle multiple sockets.
The ClientConnectionHandler will own the Socket, it doesn't need to be IDisposable.

Reply:
We may need to keep the handler objects in a collection. Though I expect the on-going async operations will keep them rooted also.